<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Train/Test Splits

_Authors: Joseph Nelson (DC), Kevin Markham (DC)_

In [17]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns

In [18]:
%matplotlib inline

In [19]:
ames_df = pd.read_csv('../assets/data/ames_train.csv')
X = ames_df.drop('SalePrice', axis='columns')
y = ames_df.loc[:, 'SalePrice']

## Simple Train/Test Splits

So far in this course we have used *simple train/test splits*:

1. Split the data set into two pieces: a **training set** and a **testing set**.
2. Fit the model on the **training set**.
3. Evaluate the model on the **testing set**.

A common rule of thumb that sklearn follows by default is to set aside 25% of your data set for testing.

### Understanding the `train_test_split` Function

In [20]:
# Do a train/test split on the Ames housing data
from sklearn.model_selection import train_test_split

X_train, X_test,y_train,y_test = train_test_split(X,y)

In [21]:
# Confirm that X_train and X_test comes from splitting X by row
print('X shape before splitting ', X.shape)
print('X shape after splitting ', X_train.shape, X_test.shape)

X shape before splitting  (1460, 80)
X shape after splitting  (1095, 80) (365, 80)


In [22]:
# Confirm that y_train and y_test comes from splitting y by row
print('y shape before splitting ', y.shape)
print('y shape after splitting ', y_train.shape, y_test.shape)

y shape before splitting  (1460,)
y shape after splitting  (1095,) (365,)


![train_test_split](../assets/images/train_test_split.png)

### Understanding the `random_state` Parameter

The `random_state` is a pseudo-random number that allows us to reproduce our results every time we run them. However, it makes it impossible to predict what are exact results will be if we chose a new `random_state`.

`random_state` is very useful for testing that your model was made correctly since it provides you with the same split each time. However, make sure you remove it if you are testing for model variability!

In [23]:
# WITHOUT a random_state parameter:
#  (If you run this code several times, you get different results!)
X_train, X_test, y_train, y_test = train_test_split(X, y)

# Print the first row of X_Train.
print(X_train.head(1))

        Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
1399  1400          50       RL         51.0     6171   Pave   NaN      Reg   

     LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence  \
1399         Lvl    AllPub  ...           0        0    NaN  MnPrv   

     MiscFeature MiscVal MoSold  YrSold  SaleType  SaleCondition  
1399         NaN       0     10    2009        WD         Normal  

[1 rows x 80 columns]


In [24]:
# WITH a random_state parameter:
#  (Same split every time! Note you can change the random state to any integer.)
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

# Print the first row of each X_train.
print(X_train.head(1))

   Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
6   7          20       RL         75.0    10084   Pave   NaN      Reg   

  LandContour Utilities  ... ScreenPorch PoolArea PoolQC Fence MiscFeature  \
6         Lvl    AllPub  ...           0        0    NaN   NaN         NaN   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  
6       0      8    2007        WD         Normal  

[1 rows x 80 columns]


In [25]:
# Load the Boston housing dataset from sklearn and print its documentation
from sklearn.datasets import load_boston

boston = load_boston()
print(boston.DESCR)

X = pd.DataFrame(boston.data, columns=boston.feature_names)
y = pd.DataFrame(boston.target, columns=['MEDV'])

boston = pd.concat([y, X], axis=1)

.. _boston_dataset:

Boston house prices dataset
---------------------------

**Data Set Characteristics:**  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pu

In [43]:
boston.corr()

MEDV      CRIM        ZN     INDUS      CHAS       NOX        RM  \
MEDV     1.000000 -0.388305  0.360445 -0.483725  0.175260 -0.427321  0.695360   
CRIM    -0.388305  1.000000 -0.200469  0.406583 -0.055892  0.420972 -0.219247   
ZN       0.360445 -0.200469  1.000000 -0.533828 -0.042697 -0.516604  0.311991   
INDUS   -0.483725  0.406583 -0.533828  1.000000  0.062938  0.763651 -0.391676   
CHAS     0.175260 -0.055892 -0.042697  0.062938  1.000000  0.091203  0.091251   
NOX     -0.427321  0.420972 -0.516604  0.763651  0.091203  1.000000 -0.302188   
RM       0.695360 -0.219247  0.311991 -0.391676  0.091251 -0.302188  1.000000   
AGE     -0.376955  0.352734 -0.569537  0.644779  0.086518  0.731470 -0.240265   
DIS      0.249929 -0.379670  0.664408 -0.708027 -0.099176 -0.769230  0.205246   
RAD     -0.381626  0.625505 -0.311948  0.595129 -0.007368  0.611441 -0.209847   
TAX     -0.468536  0.582764 -0.314563  0.720760 -0.035587  0.668023 -0.292048   
PTRATIO -0.507787  0.289946 -0.391679  0.383248 -0.121515  0.188933 -0.355501   
B        0.333461 -0.385064  0.175520 -0.356977  0.048788 -0.380051  0.128069   
LSTAT   -0.737663  0.455621 -0.412995  0.603800 -0.053929  0.590879 -0.613808   

              AGE       DIS       RAD       TAX   PTRATIO         B     LSTAT  
MEDV    -0.376955  0.249929 -0.381626 -0.468536 -0.507787  0.333461 -0.737663  
CRIM     0.352734 -0.379670  0.625505  0.582764  0.289946 -0.385064  0.455621  
ZN      -0.569537  0.664408 -0.311948 -0.314563 -0.391679  0.175520 -0.412995  
INDUS    0.644779 -0.708027  0.595129  0.720760  0.383248 -0.356977  0.603800  
CHAS     0.086518 -0.099176 -0.007368 -0.035587 -0.121515  0.048788 -0.053929  
NOX      0.731470 -0.769230  0.611441  0.668023  0.188933 -0.380051  0.590879  
RM      -0.240265  0.205246 -0.209847 -0.292048 -0.355501  0.128069 -0.613808  
AGE      1.000000 -0.747881  0.456022  0.506456  0.261515 -0.273534  0.602339  
DIS     -0.747881  1.000000 -0.494588 -0.534432 -0.232471  0.291512 -0.496996  
RAD      0.456022 -0.494588  1.000000  0.910228  0.464741 -0.444413  0.488676  
TAX      0.506456 -0.534432  0.910228  1.000000  0.460853 -0.441808  0.543993  
PTRATIO  0.261515 -0.232471  0.464741  0.460853  1.000000 -0.177383  0.374044  
B       -0.273534  0.291512 -0.444413 -0.441808 -0.177383  1.000000 -0.366087  
LSTAT    0.602339 -0.496996  0.488676  0.543993  0.374044 -0.366087  1.000000

**Exercise.**

- Make `X` a DataFrame with columns "AGE" and "RM" from the Boston housing dataset, and make `y` a series with column "MEDV."

In [28]:
X = boston.loc[:,['AGE', 'RM']]
y = boston.loc[:,'MEDV']

- Split `X` and `y` from the previosu step into a training set and a test state, setting a random state for reproducibility.

In [30]:
X_train, X_test, y_train, y_test = train_test_split(X,y, random_state = 10)

- Train a linear regression model on the training set.

In [32]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()
lr.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

- Print mean squared error (not R^2) for both the training set and the test set. You can use the `mean_squared_error` function imported below. Consult the printed documentation if you are not sure how to use it.

In [33]:
from sklearn.metrics import mean_squared_error

In [34]:
mean_squared_error?

Signature:
mean_squared_error(
    y_true,
    y_pred,
    sample_weight=None,
    multioutput='uniform_average',
)
Docstring:
Mean squared error regression loss

Read more in the :ref:`User Guide <mean_squared_error>`.

Parameters
----------
y_true : array-like of shape = (n_samples) or (n_samples, n_outputs)
    Ground truth (correct) target values.

y_pred : array-like of shape = (n_samples) or (n_samples, n_outputs)
    Estimated target values.

sample_weight : array-like of shape = (n_samples), optional
    Sample weights.

multioutput : string in ['raw_values', 'uniform_average']
    or array-like of shape (n_outputs)
    Defines aggregating of multiple output values.
    Array-like value defines weights used to average errors.

    'raw_values' :
        Returns a full set of errors in case of multioutput input.

    'uniform_average' :
        Errors of all outputs are averaged with uniform weight.

Returns
-------
loss : float or ndarray of floats
    A non-negative floating p

In [35]:
y_pred = lr.predict(X_test)
mean_squared_error(y_test, y_pred)

47.776240601175154

- Write down the equation that characterizes your fitted model. (*Hint*: use the relevant methods to print the model's intercept and coefficients, and then insert those parameters into the relevant linear regression equation.)

In [39]:
list(zip(['AGE', 'RM'],lr.coef_))
lr.intercept_

-22.471287589183476

$y = -22.47 -0.08*AGE + 7.99*RM$

- **BONUS:** See if you can get a better score on the test set by using a different set of features.

In [47]:
X = boston.loc[:,['TAX', 'RM', 'LSTAT']]
y = boston.loc[:,'MEDV']
X_train, X_test, y_train, y_test = train_test_split(X,y, random_state = 10)
lr = LinearRegression()
lr.fit(X_train, y_train)
y_pred = lr.predict(X_test)
mean_squared_error(y_test, y_pred)

39.47827319768052

## $K$-Fold Cross-Validation

You might have noticed that if you repeat the process of doing a train/test split and evaluating performance on the test set without setting the random state, you do not get the same score every time. If your dataset is not large or if it contains outliers, then the scores you get might vary a lot.

In [48]:
# Run this cell repeatedly to see how the scores can vary.
feature_cols = ['GarageCars',
                'GrLivArea',
                'OverallQual'
               ]
X = ames_df.loc[:, feature_cols]
y = ames_df.loc[:, 'SalePrice']

X_train, X_test, y_train, y_test = train_test_split(X, y)

lr = LinearRegression()

lr.fit(X_train, y_train)

print(lr.score(X_train, y_train))
print(lr.score(X_test, y_test))

0.7300796660149407
0.7522164587874405


$K$-fold cross validation helps address this problem by doing multiple train/test splits and averaging the results. For instance, five-fold cross validation consists of doing five train/test splits, each time holding out a different one-fifth of the rows, and averaging the scores over those test sets.

![](../assets/images/cross_validation_diagram.png)

You would generally use $k$-folds cross validation to develop your model and then retrain that model on the entire dataset.

- **Pro:** $K$-fold cross-validation provides more reliable estimates of model performance than a simple train/test split, especially for small datasets.
- **Con:** $K$-fold cross-validation takes longer to run because it requires training and running the model $k$ times instead of just once.

###  Varying K

Increasing $k$ augments both the advantages and disadvantages of $k$-fold cross-validation over a simple train/test split.

The extreme case of k-fold cross-validation is leave-one-out cross-validation, where you hold out just one data point at a time. This approach provides the best estimates of model perforamnce, but is requires training the model once for every row in the dataset.

A more common approach is to set $k$ somewhere between 5 and 10.

### Cross-Validation With the Boston Data

Scikit-learn provides a single function `cross_val_score` that does all of the work of doing cross-validation splits, training the model on each training set, and scoring it on the corresponding test set.

In [49]:
from sklearn.model_selection import cross_val_score

cross_val_score?

Signature:
cross_val_score(
    estimator,
    X,
    y=None,
    groups=None,
    scoring=None,
    cv='warn',
    n_jobs=None,
    verbose=0,
    fit_params=None,
    pre_dispatch='2*n_jobs',
    error_score='raise-deprecating',
)
Docstring:
Evaluate a score by cross-validation

Read more in the :ref:`User Guide <cross_validation>`.

Parameters
----------
estimator : estimator object implementing 'fit'
    The object to use to fit the data.

X : array-like
    The data to fit. Can be for example a list, or an array.

y : array-like, optional, default: None
    The target variable to try to predict in the case of
    supervised learning.

groups : array-like, with shape (n_samples,), optional
    Group labels for the samples used while splitting the dataset into
    train/test set.

scoring : string, callable or None, optional, default: None
    A string (see model evaluation documentation) or
    a scorer callable object / function with signature
    ``scorer(estimator, X, y)``.

cv 

In [54]:
ames_df = pd.read_csv('../assets/data/ames_train.csv')
ames_df.columns

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

In [56]:
# Use `cross_val_score` to do five-fold cross-validation on the Ames dataset
lr = LinearRegression()

feature_cols = ['GarageCars',
                'GrLivArea',
                'OverallQual'
               ]
X = ames_df.loc[:, feature_cols]
y = ames_df.loc[:, 'SalePrice']

scores = cross_val_score(estimator=lr, X=X, y=y, cv=5)
scores.mean()

0.7349459032805817

`cross_val_score` allows you to pass a cross-validation splitter object to the `cv` argument instead of an integer for greater control over how the splits are done. For instance, it is often a good idea to shuffle your rows before doing cross-validation in case something changes about the data as you move down the DataFrame.

In [68]:
# Use a K-fold object with `shuffle=True` to ensure that results aren't
# biased by ordering effects.
from sklearn.model_selection import KFold
kf=KFold(n_splits=5,shuffle=True)
scores = cross_val_score(estimator=lr,X=X,y=y, cv=kf)
scores.mean()

0.7298181315351785

Unfortunately, `cross_val_score` does not allow us to assess whether a model is overfitting or underfitting because it returns only test-set scores. To get training set scores as well, we need to write our own cross-validation scoring loop.

In [77]:
# Write our own cross-validation loop.
# Run this cell a few times to see how much more stable the scores are
# than they were with a simple train/test split.
train_scores = []
test_scores = []
kf =KFold(n_splits=5,shuffle=True)
for train_rows,test_rows in kf.split(X,y):
    X_train = X.iloc[train_rows,:]
    y_train = y.iloc[train_rows]
    X_test = X.iloc[test_rows,:]
    y_test = y.iloc[test_rows]
    
    lr=LinearRegression()
    lr.fit(X_train,y_train)
    train_scores.append(lr.score(X_train,y_train))
    test_scores.append(lr.score(X_test,y_test))
    
print(np.array(train_scores).mean())
print(np.array(test_scores).mean())

0.7393411982974599
0.7358577811805931


## Train/Test Splits for Time-Series Data

In time-series modeling, you are predicting future values of a variable such as a stock price from past values of that variable (perhaps along with other inputs).

The point of a train-test split is to simulate what will happen when you deploy the model. As a result, **in a useful train-test split for time series data, all of the training data comes from before all of the test data.** After all, when the model is deployed it will be used to predict future observations based on past observations.

There are many decisions to make in doing train/test splits for time series models (sometimes called *backtesting*:

- How much of a "burn-in period" will you allow before you start scoring the model?
- Will you retrain the model periodically after you start scoring it? If so, when? On all past data, or on a "rolling window" of just the most recent data?
- When you make a prediction for scoring, will you predict just one time point into the future, or multiple time points? Will you use all past data as inputs to the prediction, or just a rolling window of the most recent data?

The key principle to keep in mind in all of these decisions is that **the purpose of applying the model to test data is to simulate what will happen when you deploy it**. As a result, you should make test conditions mimic what will happen in production as much as is feasible, erring on the side of making the test more difficult so that if anything you underpromise on model performance.

## Summary

- $K$-folds cross-validation consists of doing $k$ train/test splits with disjoint test sets and averaging the resulting test-set metrics. It takes longer to run than a simple train/test split but provides more reliable estimates of test-set performance.
- Doing proper train/test splits for time-series models is tricky. You need to make sure that all of the training data comes from earlier in time than all of the test-set data and in general to simulate how the model will be used in production.